services/drupal/web/modules/custom/epa_cloudwatch/src/Logger/CloudWatch.php

```diff
@@ -0,0 +1,350 @@
+    // To save space in log messages (CloudWatch Logs imposes a maximum payload size),
+    // remove any backtraces.
+    unset($context['backtrace']);
+
+    // Convert the message timestamp from seconds to milliseconds as required by AWS.
+    $timestamp = $context['timestamp'] * 1000;
+
+    // Create a JSON message object. The values here are based on what the dblog module
+    // records.
+    $payload = json_encode([
+      'uid' => $context['uid'],
+      'type' => mb_substr($context['channel'], 0, 64),
+      'message' => $message,
+      'variables' => $this->parser->parseMessagePlaceholders($message, $context),
+      'severity' => $level,
+      'link' => $context['link'],
+      'location' => $context['request_uri'],
+      'referer' => $context['referer'],
+      'hostname' => mb_substr($context['ip'], 0, 128),
+      'timestamp' => $context['timestamp'],
+    ]);
+
+    $this->putLogMessage($payload, $timestamp);
+  }
+
+  /**
+   * Determine the CloudWatch log stream from configuration.
+   *
+   * @return boolean TRUE if a log stream was set in configuration and FALSE otherwise
+   */
+  protected function setLogStreamFromConfig() {
+    $logStream = $this->config->get('log_stream');
+    if (empty($logStream)) {
+      return FALSE;
+    }
+
+    $this->logStream = $logStream;
+    return TRUE;
+  }
+
+  /**
+   * Determine the CloudWatch log stream from ECS container metadata.
+   *
+   * This relies on the presence of the $ECS_CONTAINER_METADATA_URI_V4 environment
+   * variable and container metadata being available at the referenced endpoint.
+   *
+   * @see https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v4.html
+   *
+   * @return boolean TRUE if a log stream name could be determined and FALSE otherwise
+   */
+  protected function setLogStreamFromContainerMetadata() {
+    // First, find the endpoint in the container environment
+    $endpoint = getenv('ECS_CONTAINER_METADATA_URI_V4');
+    if (empty($endpoint)) {
+      return FALSE;
+    }
+
+    // Next, load the value from the endpoint
+    $json = file_get_contents($endpoint);
+    if (empty($json)) {
+      return FALSE;
+    }
+
+    // Use the Docker container ID as the stream name, same as the ECS logs agent.
+    $metadata = json_decode($json, FALSE, 0, JSON_THROW_ON_ERROR);
+
+    $this->logStream = $metadata->DockerId;
+    return TRUE;
+  }
+
+  /**
+   * Determines the log stream name to use.
+   *
+   * @return string The CloudWatch log stream name
+   */
+  protected function getLogStream() {
+    if (isset($this->logStream)) {
+      return $this->logStream;
+    }
+
+    if ($this->setLogStreamFromConfig()) {
+      return $this->logStream;
+    }
+
+    if ($this->setLogStreamFromContainerMetadata()) {
+      return $this->logStream;
+    }
+
+    // If we can't determine a log stream name, throw an exception.
+    throw new \Exception('Unable to determine log stream name.');
```

Comment:
Coding standards [_prefer_](https://www.drupal.org/docs/develop/coding-standards/php-exceptions#subclass) a subclassed Exception, but I can't think of a scenario where we would ever catch this exception or do anything other than exit since the log stream name is required. I'm fine with keeping as-is.